File: framework/src/main/java/me/insidezhou/southernquiet/debounce/DefaultDebouncerProvider.java
package me.insidezhou.southernquiet.debounce;

import me.insidezhou.southernquiet.FrameworkAutoConfiguration;
import me.insidezhou.southernquiet.logging.SouthernQuietLogger;
import me.insidezhou.southernquiet.logging.SouthernQuietLoggerFactory;
import me.insidezhou.southernquiet.util.Pair;
import me.insidezhou.southernquiet.util.Tuple;
import org.aopalliance.intercept.MethodInvocation;
import org.springframework.beans.factory.DisposableBean;
import org.springframework.util.StringUtils;

import java.lang.reflect.Method;
import java.time.Duration;
import java.util.Arrays;
import java.util.Objects;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicLong;
import java.util.stream.IntStream;

@SuppressWarnings("DuplicatedCode")
public class DefaultDebouncerProvider implements DebouncerProvider, DisposableBean {
    private final static SouthernQuietLogger log = SouthernQuietLoggerFactory.getLogger(DefaultDebouncerProvider.class);

    private final ConcurrentMap<String, Pair<Debouncer, MethodInvocation>> debouncerAndInvocations = new ConcurrentHashMap<>();
    private final ConcurrentLinkedQueue<Tuple<String, Debouncer, MethodInvocation>> pendingInvocations = new ConcurrentLinkedQueue<>();

    private final Duration reportDuration;
    private long reportTimer = System.currentTimeMillis();

    private long checkCounter = 0;
    private final AtomicLong workCounter = new AtomicLong(0);

    private final ScheduledExecutorService executorService;

    public DefaultDebouncerProvider(FrameworkAutoConfiguration.DebounceProperties properties) {
        this.reportDuration = properties.getReportDuration();
        executorService = Executors.newScheduledThreadPool(properties.getWorkerCount() + 1);

        executorService.scheduleAtFixedRate(this::checkDebouncer, 1, 1, TimeUnit.MILLISECONDS);

        IntStream.range(0, properties.getWorkerCount()).forEach(i -> {
            executorService.scheduleAtFixedRate(this::invokeDebouncer, 1, 1, TimeUnit.MILLISECONDS);
        });
    }

    @Override
    public Debouncer getDebouncer(MethodInvocation invocation, long waitFor, long maxWaitFor, String debouncerName) {
        Object bean = invocation.getThis();
        Method method = invocation.getMethod();

        if (StringUtils.isEmpty(debouncerName)) {
            debouncerName = bean.getClass().getName() + "#" + method.getName() + "_" + waitFor + "_" + maxWaitFor;
        }

        Pair<Debouncer, MethodInvocation> pair = debouncerAndInvocations.computeIfAbsent(debouncerName, (nm) -> new Pair<>(new DefaultDebouncer(waitFor, maxWaitFor), invocation));
        pair.setSecond(invocation);
        debouncerAndInvocations.put(debouncerName, pair);

        return pair.getFirst();
    }

    private void checkDebouncer() {
        Arrays.stream(debouncerAndInvocations.keySet().toArray(new String[0]))
            .map(name -> {
                Pair<Debouncer, MethodInvocation> pair = debouncerAndInvocations.get(name);
                if (null == pair) return null;

                Debouncer debouncer = pair.getFirst();
                MethodInvocation invocation = pair.getSecond();

                return new Tuple<>(name, debouncer, invocation);
            })
            .filter(Objects::nonNull)
            .filter(tuple -> tuple.getSecond().isStable())
            .forEach(tuple -> {
                ++checkCounter;
                pendingInvocations.add(tuple);
                debouncerAndInvocations.remove(tuple.getFirst());
            });

        long now = System.currentTimeMillis();
        Duration interval = Duration.ofMillis(now - reportTimer);
        if (interval.compareTo(reportDuration) >= 0) {
            long work = workCounter.getAndSet(0);

            log.message("debouncer计数器")
                .context("check", checkCounter)
                .context("work", work)
                .context("pending", pendingInvocations.size())
                .context("interval", interval)
                .debug();

            checkCounter = 0;
            reportTimer = now;
        }
    }

    private void invokeDebouncer() {
        Tuple<String, Debouncer, MethodInvocation> tuple = pendingInvocations.poll();
        while (null != tuple) {
            String name = tuple.getFirst();
            MethodInvocation invocation = tuple.getThird();

            try {
                invocation.proceed();
            }
            catch (Throwable throwable) {
                log.message("施加了去抖动的方法执行失败")
                    .context("debouncer", name)
                    .exception(throwable)
                    .error();
            }

            workCounter.incrementAndGet();
            tuple = pendingInvocations.poll();
        }
    }

    @Override
    public void destroy() {
        executorService.shutdown();
    }
}

File: framework/src/main/kotlin/me/insidezhou/southernquiet/debounce/DefaultDebouncerProvider.kt
package me.insidezhou.southernquiet.debounce

import kotlinx.coroutines.*
import me.insidezhou.southernquiet.FrameworkAutoConfiguration.DebounceProperties
import me.insidezhou.southernquiet.logging.SouthernQuietLoggerFactory
import me.insidezhou.southernquiet.util.Metadata
import org.aopalliance.intercept.MethodInvocation
import org.springframework.beans.factory.DisposableBean
import org.springframework.util.StringUtils
import java.time.Duration
import java.util.*
import java.util.concurrent.*
import java.util.concurrent.atomic.AtomicLong
import kotlin.math.round

@Suppress("MemberVisibilityCanBePrivate")
open class DefaultDebouncerProvider(properties: DebounceProperties, val dispatcher: CoroutineDispatcher) : DebouncerProvider, DisposableBean {
    constructor(properties: DebounceProperties, metadata: Metadata) : this(
        properties,
        ThreadPoolExecutor(
            metadata.coreNumber,
            Int.MAX_VALUE,
            maxOf(1L, round(metadata.coreNumber / 20.0).toLong()),
            TimeUnit.SECONDS,
            ArrayBlockingQueue<Runnable>(metadata.coreNumber * 30)
        ).asCoroutineDispatcher()
    )

    private val debouncerAndInvocations = ConcurrentHashMap<String, DebouncerMetadata>()
    private val pendingInvocations = LinkedList<DebouncerMetadata>()
    private val reportDuration: Duration = properties.reportDuration
    private var reportTimer = System.currentTimeMillis()

    private val pendingCounter = AtomicLong(0)
    private val putCounter = AtomicLong(0)
    private val doneCounter = AtomicLong(0)

    private val scheduledFuture = Executors.newSingleThreadScheduledExecutor().scheduleAtFixedRate(
        {
            checkDebouncer()
            workDebouncer()
        },
        1,
        1,
        TimeUnit.MILLISECONDS)

    private val workCoroutineScope = CoroutineScope(dispatcher)

    override fun getDebouncer(invocation: MethodInvocation, waitFor: Long, maxWaitFor: Long, name: String, executionTimeout: Long): Debouncer {
        var debouncerName = name

        val bean = invocation.getThis()
        val method = invocation.method
        if (StringUtils.isEmpty(debouncerName)) {
            debouncerName = bean.javaClass.name + "#" + method.name + "_" + waitFor + "_" + maxWaitFor
        }

        val metadata = debouncerAndInvocations.getOrPut(debouncerName) {
            log.message("准备生成debouncer")
                .context("name", debouncerName)
                .context("class", bean.javaClass.simpleName)
                .context("method", method.name)
                .context("waitFor", waitFor)
                .context("maxWaitFor", maxWaitFor)
                .context("timeout", executionTimeout)
                .debug()

            putCounter.incrementAndGet()

            DebouncerMetadata(debouncerName, DefaultDebouncer(waitFor, maxWaitFor), invocation, executionTimeout)
        }

        return metadata.debouncer
    }

    private fun checkDebouncer() {
        debouncerAndInvocations
            .filter { it.value.debouncer.isStable }
            .forEach {
                pendingInvocations.add(it.value)
                debouncerAndInvocations.remove(it.key)
            }

        pendingCounter.addAndGet(pendingInvocations.size.toLong())

        val now = System.currentTimeMillis()
        val interval = Duration.ofMillis(now - reportTimer)
        if (interval >= reportDuration) {
            val put = putCounter.getAndSet(0)
            val done = doneCounter.getAndSet(0)
            val unstable = debouncerAndInvocations.size

            reportTimer = now

            val executor = dispatcher.asExecutor()

            log.message("debouncer计数器")
                .context {
                    it["put"] = put
                    it["done"] = done
                    it["interval"] = interval
                    it["unstable"] = unstable
                    it["pending"] = pendingCounter.get()

                    if (executor is ThreadPoolExecutor) {
                        it["working"] = executor.activeCount
                        it["pool"] = executor.poolSize
                        it["queue"] = executor.queue.size
                    }
                }
                .trace()
        }
    }

    private fun workDebouncer() {
        do {
            val metadata = pendingInvocations.poll() ?: return

            workCoroutineScope.launch {
                val invocation = metadata.invocation
                val timeout = metadata.executionTimeout

                try {
                    if (timeout > 0) {
                        val result = withTimeoutOrNull(timeout) {
                            invocation.proceed()
                            "DONE"
                        }

                        if (null == result) {
                            onWorkTimeout(metadata)
                        }
                    }
                    else {
                        invocation.proceed()
                    }
                }
                catch (throwable: Throwable) {
                    onWorkException(throwable, metadata)
                }
                finally {
                    pendingCounter.decrementAndGet()
                    doneCounter.incrementAndGet()
                }
            }
        }
        while (true)
    }

    protected open fun onWorkException(throwable: Throwable, metadata: DebouncerMetadata) {
        log.message("施加了去抖动的方法执行失败")
            .context("debouncer", metadata.name)
            .exception(throwable)
            .error()
    }

    protected open fun onWorkTimeout(metadata: DebouncerMetadata) {
        log.message("施加了去抖动的方法执行超时")
            .context("debouncer", metadata.name)
            .warn()
    }

    override fun destroy() {
        scheduledFuture.cancel(true)
    }

    companion object {
        private val log = SouthernQuietLoggerFactory.getLogger(DefaultDebouncerProvider::class.java)
    }
}

class DebouncerMetadata(val name: String, val debouncer: Debouncer, val invocation: MethodInvocation, val executionTimeout: Long)
